x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/highlight/SemanticTextHighlighter.java

```diff
@@ -0,0 +1,208 @@
+    public static final String NAME = "semantic";
+
+    private record OffsetAndScore(int offset, float score) {}
+
+    @Override
+    public boolean canHighlight(MappedFieldType fieldType) {
+        if (fieldType instanceof SemanticTextFieldMapper.SemanticTextFieldType) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public HighlightField highlight(FieldHighlightContext fieldContext) throws IOException {
+        SemanticTextFieldMapper.SemanticTextFieldType fieldType = (SemanticTextFieldMapper.SemanticTextFieldType) fieldContext.fieldType;
+        if (fieldType.getEmbeddingsField() == null) {
+            // nothing indexed yet
+            return null;
+        }
+
+        final List<Query> queries = switch (fieldType.getModelSettings().taskType()) {
+            case SPARSE_EMBEDDING -> extractSparseVectorQueries(
+                (SparseVectorFieldType) fieldType.getEmbeddingsField().fieldType(),
+                fieldContext.query
+            );
+            case TEXT_EMBEDDING -> extractDenseVectorQueries(
+                (DenseVectorFieldType) fieldType.getEmbeddingsField().fieldType(),
+                fieldContext.query
+            );
+            default -> throw new IllegalStateException(
+                "Wrong task type for a semantic text field, got [" + fieldType.getModelSettings().taskType().name() + "]"
+            );
+        };
+        if (queries.isEmpty()) {
+            // nothing to highlight
+            return null;
+        }
+
+        int numberOfFragments = fieldContext.field.fieldOptions().numberOfFragments() <= 0
+            ? 1 // we return the best fragment by default
+            : fieldContext.field.fieldOptions().numberOfFragments();
+
+        List<OffsetAndScore> chunks = extractOffsetAndScores(
+            fieldContext.context.getSearchExecutionContext(),
+            fieldContext.hitContext.reader(),
+            fieldType,
+            fieldContext.hitContext.docId(),
+            queries
+        );
+        if (chunks.size() == 0) {
+            return null;
+        }
+
+        chunks.sort(Comparator.comparingDouble(OffsetAndScore::score).reversed());
+        int size = Math.min(chunks.size(), numberOfFragments);
+        if (fieldContext.field.fieldOptions().scoreOrdered() == false) {
+            chunks = chunks.subList(0, size);
+            chunks.sort(Comparator.comparingInt(c -> c.offset));
+        }
+        Text[] snippets = new Text[size];
+        List<Map<?, ?>> nestedSources = XContentMapValues.extractNestedSources(
+            fieldType.getChunksField().fullPath(),
+            fieldContext.hitContext.source().source()
+        );
+        for (int i = 0; i < size; i++) {
+            var chunk = chunks.get(i);
+            if (nestedSources.size() <= chunk.offset) {
+                throw new IllegalStateException("Invalid content for field [" + fieldType.name() + "]");
```

Comment:
Could we make the error message a bit more descriptive here?

Reply:
++ maybe include size and offset?

Reply:
Done in https://github.com/elastic/elasticsearch/pull/118064/commits/4639c31252473aebb45641d122eb9c24fd341d3c